File: src/main/java/dev/jlo/gamemodes/paper/MatchSession.java
package dev.jlo.gamemodes.paper;

import dev.jlo.gamemodes.domain.common.MatchId;
import dev.jlo.gamemodes.domain.common.MatchPhase;
import dev.jlo.gamemodes.domain.common.MatchResult;
import dev.jlo.gamemodes.domain.common.Team;
import dev.jlo.gamemodes.domain.opr.OprConfig;
import dev.jlo.gamemodes.domain.opr.OprMatch;
import dev.jlo.gamemodes.domain.opr.OutpostId;
import dev.jlo.gamemodes.domain.siege.SiegeConfig;
import dev.jlo.gamemodes.domain.siege.SiegeMatch;
import dev.jlo.gamemodes.domain.siege.RallyPoint;
import dev.jlo.gamemodes.domain.siege.SiegeWeapon;
import dev.jlo.gamemodes.domain.siege.Structure;
import org.bukkit.Location;
import org.bukkit.entity.Player;

import java.time.Clock;
import java.time.Instant;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.UUID;

/** Owns one arena reservation and adapts either deterministic rules engine to the Paper layer. */
public class MatchSession {
    private final Mode mode;
    private final String arenaId;
    private final String id;
    private final MatchQueue queue;
    private final SiegeConfig configuredSiege;
    private final Map<UUID, Team> teams = new LinkedHashMap<>();
    private final OprMatch opr;
    private final SiegeMatch siege;
    private final Clock clock;
    private Instant preparationStartedAt;
    private Instant battleStartedAt;

    public MatchSession(Mode mode, String arenaId, int capacityPerTeam, int quorumPerTeam) {
        this(mode, arenaId, capacityPerTeam, quorumPerTeam, Clock.systemUTC(),
                new OprConfig(capacityPerTeam, quorumPerTeam), new SiegeConfig(quorumPerTeam));
    }

    public MatchSession(Mode mode, String arenaId, int capacityPerTeam, int quorumPerTeam, Clock clock) {
        this(mode, arenaId, capacityPerTeam, quorumPerTeam, clock,
                new OprConfig(capacityPerTeam, quorumPerTeam), new SiegeConfig(quorumPerTeam));
    }

    public MatchSession(Mode mode, String arenaId, int capacityPerTeam, int quorumPerTeam,
                        Clock clock, OprConfig oprConfig, SiegeConfig siegeConfig) {
        this.mode = mode;
        this.arenaId = arenaId;
        this.clock = clock;
        this.id = mode + ":" + arenaId + ":" + UUID.randomUUID();
        this.queue = new MatchQueue(mode, capacityPerTeam, quorumPerTeam);
        this.configuredSiege = siegeConfig;
        Instant now = clock.instant();
        this.opr = mode == Mode.OPR ? new OprMatch(oprConfig, now) : null;
        this.siege = mode == Mode.SIEGE ? new SiegeMatch(new MatchId(id), siegeConfig, now) : null;
    }

    public Mode getMode() {
        return mode;
    }

    public String getArenaId() {
        return arenaId;
    }

    public String getId() {
        return id;
    }

    public MatchPhase getPhase() {
        return opr != null ? opr.getLifecycle().getPhase() : siege.getLifecycle().getPhase();
    }

    public MatchResult getResult() {
        return opr != null ? opr.getResult() : siege.getResult();
    }

    public synchronized TeamAssignment join(UUID player) {
        return join(player, null);
    }

    public synchronized TeamAssignment join(UUID player, TeamAssignment preferred) {
        Team team = queue.join(player, preferred == null ? null : preferred == TeamAssignment.A ? Team.A : Team.B);
        teams.put(player, team);
        if (opr != null) {
            opr.assign(player, team);
        } else {
            siege.assign(player, team);
        }
        return team == Team.A ? TeamAssignment.A : TeamAssignment.B;
    }

    public synchronized TeamAssignment leave(UUID player) {
        Team team = queue.leave(player);
        if (team == null) return null;
        teams.remove(player);
        if (opr != null) {
            opr.leave(player);
        } else {
            siege.leave(player);
        }
        return team == Team.A ? TeamAssignment.A : TeamAssignment.B;
    }

    public synchronized TeamAssignment teamOf(UUID player) {
        Team team = queue.teamOf(player);
        return team == null ? null : team == Team.A ? TeamAssignment.A : TeamAssignment.B;
    }

    public synchronized boolean ready(UUID player) {
        if (siege != null) {
            queue.markReady(player);
            siege.ready(player);
            return true;
        }
        return queue.markReady(player);
    }

    public synchronized boolean startIfReady() {
        MatchPhase phase = getPhase();
        if (phase == MatchPhase.RESOLVING || phase == MatchPhase.CLEANUP) return false;
        if (!queue.hasQuorum()) return false;
        Instant now = clock.instant();
        if (siege != null) {
            if (phase == MatchPhase.WAITING) {
                siege.startPreparing(now);
                preparationStartedAt = now;
            }
            if (getPhase() == MatchPhase.PREPARING && queue.allReady()) {
                siege.advanceTo(now);
                if (getPhase() == MatchPhase.PREPARING && configuredSiege.getPreparation().isZero()) {
                    siege.advanceTo(now.plusNanos(1));
                }
            }
        } else {
            if (phase == MatchPhase.WAITING) {
                opr.beginPreparation(now);
                preparationStartedAt = now;
            }
            if (getPhase() == MatchPhase.PREPARING) opr.start(now);
        }
        if (getPhase() == MatchPhase.ACTIVE && battleStartedAt == null) battleStartedAt = now;
        return getPhase() == MatchPhase.ACTIVE;
    }

    public synchronized void advance() {
        Instant now = clock.instant();
        if (opr != null) opr.advanceTo(now); else siege.advanceTo(now);
    }

    public void advanceAt(Instant at) {
        if (opr != null) opr.advanceTo(at); else siege.advanceTo(at);
    }

    public boolean captureRally(RallyPoint point, UUID player, Instant at) {
        return siege != null && siege.captureRally(point, player, at);
    }

    public boolean damageGate(UUID player, int amount, Instant at) {
        return siege != null && siege.damageStructure(Structure.GATE, SiegeWeapon.CANNON, player, amount, at);
    }

    public boolean beginClaim(UUID player, Instant at) {
        return siege != null && siege.beginClaim(player, at);
    }

    public boolean completeClaim(UUID player, Instant at) {
        return siege != null && siege.completeClaim(player, at);
    }

    public void enterOutpost(UUID player, OutpostId outpost) {
        if (opr != null) opr.enterOutpost(player, outpost);
    }

    public void disconnect(UUID player) {
        if (opr != null) opr.disconnect(player, clock.instant());
        else if (siege != null) siege.disconnect(player);
    }

    public boolean recordCombat(UUID attacker, UUID victim) {
        return opr != null && opr.recordCombat(attacker, victim, clock.instant());
    }

    public boolean playerDied(UUID player) {
        return playerDied(player, null);
    }

    public boolean playerDied(UUID player, UUID killer) {
        if (getPhase() != MatchPhase.ACTIVE || teamOf(player) == null) return false;
        TeamAssignment killerTeam = killer == null ? null : teamOf(killer);
        TeamAssignment playerTeam = teamOf(player);
        UUID sameSessionEnemy = killerTeam != null && killerTeam != playerTeam ? killer : null;
        if (opr != null) return opr.recordDeath(player, sameSessionEnemy, clock.instant());
        siege.recordDeath(player, sameSessionEnemy);
        return sameSessionEnemy != null;
    }

    public Location respawnLocation(Player player) {
        return null;
    }

    public void reconcileObjective(UUID player, Location location) {
        if (opr == null || getPhase() != MatchPhase.ACTIVE) return;
        OutpostId[] entries = OutpostId.values();
        int index = Math.floorMod(location.getBlockX() + location.getBlockZ(), entries.length);
        OutpostId objective = entries[index];
        for (OutpostId outpost : entries) {
            if (outpost != objective) opr.leaveOutpost(player, outpost);
        }
        opr.enterOutpost(player, objective);
    }

    public void recordKill(UUID killer, UUID victim) {
        if (opr != null) opr.recordKill(killer, victim, clock.instant());
    }

    public synchronized void cleanup() {
        if (opr != null) {
            if (getPhase() != MatchPhase.WAITING) opr.resolveAndCleanup(clock.instant());
        } else if (getPhase() != MatchPhase.WAITING) {
            siege.abort();
            siege.cleanup();
        }
        for (UUID player : teams.keySet().toArray(UUID[]::new)) queue.leave(player);
        teams.clear();
    }
}

File: src/main/kotlin/dev/jlo/gamemodes/paper/MatchSession.kt
package dev.jlo.gamemodes.paper

import dev.jlo.gamemodes.domain.common.MatchPhase
import dev.jlo.gamemodes.domain.common.MatchResult
import dev.jlo.gamemodes.domain.common.Team
import dev.jlo.gamemodes.domain.opr.OprConfig
import dev.jlo.gamemodes.domain.opr.OprMatch
import dev.jlo.gamemodes.domain.siege.SiegeConfig
import dev.jlo.gamemodes.domain.siege.SiegeMatch
import java.time.Clock
import java.util.UUID

@Suppress("unused")
enum class TeamAssignment { A, B }

/** Owns one arena reservation and adapts either deterministic rules engine to the Paper layer. */
class MatchSession(
    val mode: Mode,
    val arenaId: String,
    capacityPerTeam: Int,
    quorumPerTeam: Int,
    private val clock: Clock = Clock.systemUTC(),
    oprConfig: OprConfig = OprConfig(teamCapacity = capacityPerTeam, quorumPerTeam = quorumPerTeam),
    siegeConfig: SiegeConfig = SiegeConfig(quorum = quorumPerTeam)
) {
    val id: String = "$mode:$arenaId:${UUID.randomUUID()}"
    private val queue = MatchQueue(mode, capacityPerTeam, quorumPerTeam)
    private val configuredSiege = siegeConfig
    private val teams = linkedMapOf<UUID, Team>()
    private val opr = if (mode == Mode.OPR) OprMatch(oprConfig, clock.instant()) else null
    private val siege = if (mode == Mode.SIEGE) SiegeMatch(dev.jlo.gamemodes.domain.common.MatchId(id), configuredSiege, clock.instant()) else null

    private var preparationStartedAt: java.time.Instant? = null
    private var battleStartedAt: java.time.Instant? = null

    val phase: MatchPhase get() = when {
        opr != null -> opr.lifecycle.phase
        else -> siege!!.lifecycle.phase
    }

    val result: MatchResult? get() = opr?.result ?: siege?.result

    @Synchronized
    fun join(player: UUID, preferred: TeamAssignment? = null): TeamAssignment {
        val team = queue.join(player, preferred?.let { if (it == TeamAssignment.A) Team.A else Team.B })
        teams[player] = team
        if (opr != null) opr.assign(player, team) else siege!!.assign(player, team)
        return if (team == Team.A) TeamAssignment.A else TeamAssignment.B
    }

    @Synchronized
    fun leave(player: UUID): TeamAssignment? {
        val team = queue.leave(player) ?: return null
        teams.remove(player)
        if (opr != null) opr.leave(player) else siege!!.leave(player)
        return if (team == Team.A) TeamAssignment.A else TeamAssignment.B
    }

    @Synchronized
    fun teamOf(player: UUID): TeamAssignment? = queue.teamOf(player)?.let { if (it == Team.A) TeamAssignment.A else TeamAssignment.B }

    @Synchronized
    fun ready(player: UUID): Boolean = if (siege != null) {
        queue.markReady(player)
        siege!!.ready(player)
        true
    } else queue.markReady(player)

    @Synchronized
    fun startIfReady(): Boolean {
        if (phase == MatchPhase.RESOLVING || phase == MatchPhase.CLEANUP) return false
        if (!queue.hasQuorum()) return false
        val now = clock.instant()
        if (siege != null) {
            if (phase == MatchPhase.WAITING) {
                siege.startPreparing(now)
                preparationStartedAt = now
            }
            if (phase == MatchPhase.PREPARING && queue.allReady()) {
                siege.advanceTo(now)
                if (phase == MatchPhase.PREPARING && configuredSiege.preparation.isZero) siege.advanceTo(now.plusNanos(1))
            }
        } else {
            if (phase == MatchPhase.WAITING) {
                opr!!.beginPreparation(now)
                preparationStartedAt = now
            }
            if (phase == MatchPhase.PREPARING) opr!!.start(now)
        }
        if (phase == MatchPhase.ACTIVE && battleStartedAt == null) battleStartedAt = now
        return phase == MatchPhase.ACTIVE
    }

    @Synchronized
    fun advance() {
        val now = clock.instant()
        if (opr != null) opr.advanceTo(now) else siege!!.advanceTo(now)
    }

    fun advanceAt(at: java.time.Instant) {
        if (opr != null) opr.advanceTo(at) else siege!!.advanceTo(at)
    }

    fun captureRally(point: dev.jlo.gamemodes.domain.siege.RallyPoint, player: UUID, at: java.time.Instant): Boolean =
        siege?.captureRally(point, player, at) ?: false

    fun damageGate(player: UUID, amount: Int, at: java.time.Instant): Boolean =
        siege?.damageStructure(dev.jlo.gamemodes.domain.siege.Structure.GATE, dev.jlo.gamemodes.domain.siege.SiegeWeapon.CANNON, player, amount, at) ?: false

    fun beginClaim(player: UUID, at: java.time.Instant): Boolean = siege?.beginClaim(player, at) ?: false

    fun completeClaim(player: UUID, at: java.time.Instant): Boolean = siege?.completeClaim(player, at) ?: false
    fun enterOutpost(player: UUID, outpost: dev.jlo.gamemodes.domain.opr.OutpostId) {
        opr?.enterOutpost(player, outpost)
    }

    fun disconnect(player: UUID) {
        if (opr != null) {
            opr.disconnect(player, clock.instant())
        } else {
            siege?.disconnect(player)
        }
    }

    fun recordCombat(attacker: UUID, victim: UUID): Boolean =
        opr?.recordCombat(attacker, victim, clock.instant()) ?: false

    fun playerDied(player: UUID, killer: UUID? = null): Boolean {
        if (phase != MatchPhase.ACTIVE || teamOf(player) == null) return false
        val sameSessionEnemy = killer?.takeIf {
            teamOf(it) != null && teamOf(it) != teamOf(player)
        }
        return if (opr != null) {
            opr.recordDeath(player, sameSessionEnemy, clock.instant())
        } else {
            siege!!.recordDeath(player, sameSessionEnemy)
            sameSessionEnemy != null
        }
    }

    fun respawnLocation(player: org.bukkit.entity.Player): org.bukkit.Location? = null

    fun reconcileObjective(player: UUID, location: org.bukkit.Location) {
        if (opr == null || phase != MatchPhase.ACTIVE) return
        val index = Math.floorMod(location.blockX + location.blockZ, dev.jlo.gamemodes.domain.opr.OutpostId.entries.size)
        val objective = dev.jlo.gamemodes.domain.opr.OutpostId.entries[index]
        dev.jlo.gamemodes.domain.opr.OutpostId.entries.filter { it != objective }.forEach { opr.leaveOutpost(player, it) }
        opr.enterOutpost(player, objective)
    }
    fun recordKill(killer: UUID, victim: UUID) {
        opr?.recordKill(killer, victim, clock.instant())
    }

    @Synchronized
    fun cleanup() {
        if (opr != null) {
            if (phase != MatchPhase.WAITING) opr.resolveAndCleanup(clock.instant())
        } else if (phase != MatchPhase.WAITING) {
            siege!!.abort()
            siege!!.cleanup()
        }
        teams.keys.toList().forEach { queue.leave(it) }
        teams.clear()
    }
}
